Discover ux hit actions

lmcnulty opened the pull request:
Resolves #537. Changes:

- Makes result header images and titles into links to the report text on our site
- Make the subtitle containing the domain into a link to the original source
- Removes blockquotes from descriptions
  - These are usually redundant and take up vertical space.
  - In compact view, they also sometimes push the titles outside of the card
- Filters options are start out collapsed under a "filter search" `<summary>`.
  - This adds enough vertical space that the result titles appear above the fold.
  - On mobile, the modal filter dialogue is still used
  - If we want to highlight the filter options more, another possibility would be to put them in a sidebar
    - It might be better to de-emphasize the filters until we've solved some of the other UX issues (see #630).
    - I think that the "Classifications" filter is much more interesting than any of the others. Perhaps we could put it in the first row but hide the others behind the click?

## Before

The result titles are mostly below the fold, and they are not clickable.

https://user-images.githubusercontent.com/25443411/171270128-e700e2db-7751-4c4c-b542-2234638a4c6b.mp4

## After

Titles are above the fold, and header images, titles, and source domains are links.

https://user-images.githubusercontent.com/25443411/171270335-786d281f-282c-4cf5-85fb-7d5aaaedf0b8.mp4

smcgregor commented:
Design review: the filters should be expanded on page load if a filter is applied. I'll look closer at the code after that fix is in, and play with it interactively.

smcgregor commented:
Build is failing. Can you try merging staging into this branch to see if that fixes it? I plan on playing with the functionality then returning to the code review.

smcgregor commented on site/gatsby-site/src/components/discover/hitTypes/Details.js:
```diff
@@ -19,41 +20,29 @@ const StyledLabel = styled.p`
   flagReportModal,
   toggleFilterByIncidentId,
 }) {
+  console.log(item);
```
Leftover logging?

lmcnulty replied:
Good catch. I wonder if it would be worth it to add `no-console` to the eslint config for src/components and src/pages?

smcgregor commented:
I played with the deploy preview and there appears to be a bug. Replicating:

1. Open the discover application
2. Apply a filter
3. Refresh the page
4. (bug 1): the filters are not expanded and visible
5. Expand the filters
6. (bug 2): the color of the applied filter is not changed, but the number of filters applied is properly indicated

lmcnulty commented:
The "[Generate a Algolia index](https://app.netlify.com/sites/staging-aiid/deploys/62a3351f87e31b0009385fd3#L640)" test is now failing... I think that's caused by recent, unrelated changes, since #685 failed it as well.

If that's not an issue, I think this is ready for another review / merge.

smcgregor commented:
1. `Controls` should be implemented in the `Filters` component. You moved some of the functionality into `Discover` page.
2. Rather than list out every refinement to check for, you should rely on a single source of truth and pull from `import REFINEMENT_LISTS from 'components/discover/REFINEMENT_LISTS';` as is done to render the `StyledFilter` in Filters.js. ([example](https://github.com/responsible-ai-collaborative/aiid/blob/master/site/gatsby-site/src/components/discover/REFINEMENT_LISTS.js#L24)).

Please let me know if I am missing something with (1) and (2) above. Once they are addressed, please ask @cesarvarela to review and provide an explanation for how you landed on using a `click` event on first render. He may have ideas there.

lmcnulty commented:
Regarding the feedback:

1. I somewhat disagree that `Controls` belongs in `Filters`. The parts that I moved to `Discover` were `Stats`, `DisplayModeSwitch`, and `ClearFilters`. The first two are basically unrelated to the filters – they would work just as well if there were no filtering functionality. `ClearFilters` is logically filter-related, but it's displayed independent of the filters. At least for now, it's positioned in the row above the filters, and it's shown greyed-out even when the filters are not rendered. So none of those components feel like they're part of the filters component.
  
    However, I think the underlying issue you were getting at was that `discover.js` had grown too long. So I moved `Controls` to its own file rather than to adding it to `Filters`. Is that ok?

2. Thanks, I missed that the refinements were coming from there. I've made the change.

Regarding the use of the `.click()` event: After I had switched to looking at the `query` rather than the `searchState`, the filters were correctly displaying on page load when one or more was applied. However, they were displaying in a single column regardless of the viewport width. I didn't look into it carefully enough to figure out for sure why that was occurring, but hiding and then showing the filters again caused them to display correctly. I'm guessing that the responsive design depended on the js `window` object, which was unavailable in server-side rendering, so it would only display correctly when a client-side render was triggered. `.click()` seems to accomplish that, although perhaps there's a better way?

cesarvarela commented on site/gatsby-site/src/components/discover/Controls.js:
```diff
@@ -0,0 +1,72 @@
+        }
+      }
+      if (filter) {
+        document.querySelector('#expand-filters').click();
```
Direct DOM manipulation is a big no-no! Worst case use a ref, but here seems like you would want to use a state flag.

lmcnulty replied:
I've changed it to use `setExpandFilters(true)`.

smcgregor commented:
It should likely be filters are in controls which is referenced in discover. I think you did that here. :)

It looks like @cesarvarela has some suggestions on the click.

lmcnulty commented on site/gatsby-site/src/components/discover/Controls.js:
```diff
@@ -0,0 +1,72 @@
+      if (filter) {
+        document.querySelector('#expand-filters').click();
+      }
+      setFirstRender(false);
```
Done.

cesarvarela commented on site/gatsby-site/src/components/discover/Controls.js:
```diff
@@ -0,0 +1,67 @@
+  const [expandFilters, setExpandFilters] = useState(false);
+
+  useEffect(() => {
+    let filter = false;
```
No need to change this but `some` can save you some lines in the future, and the `if` is redundant.
```
const filter = REFINEMENT_LISTS.some(r => query[r.attribute]);
setExpandFilters(filter);
```

cesarvarela commented on site/gatsby-site/src/components/discover/Controls.js:
```diff
@@ -0,0 +1,67 @@
+          </ExpandFilters>
+        </Col>
+      </Row>
+      <Row className="mb-3 hiddenMobile">{expandFilters ? <Filters /> : <div></div>}</Row>
```
No need to use an empty div here: `{expandFilters && <Filters />}`,

cesarvarela reviewed (approved):
Added two nitpicks, good to merge anyway.

smcgregor commented:
I whitelisted the domain so the deploy preview should work, and I am not seeing the arrow for the filters. Any ideas?

![Screen Shot 2022-06-14 at 9 38 12 PM](https://user-images.githubusercontent.com/64780/173737872-2d243cdd-ae48-4f59-9366-9381d15276a2.png)

lmcnulty commented:
The arrow was displayed with the unicode character, so it's probably a font issue. I've switched to using an svg icon – let me know if that fixes it.
